fix: Fix issue with volume mount

peiniliu opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/couler-proj/couler/blob/master/CODE_OF_CONDUCT.md
  2. Ensure you have added or run the appropriate tests for your PR: https://github.com/couler-proj/couler/blob/master/CONTRIBUTING.md
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. 
-->

fix  [Issue #193 ](https://github.com/couler-proj/couler/issues/193#issue-854389174)

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

fix  [Issue #193 ](https://github.com/couler-proj/couler/issues/193#issue-854389174)

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

volume_test.py

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/couler-proj/couler?pullRequest=194) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/couler-proj/couler?pullRequest=194) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/couler-proj/couler?pullRequest=194) it.</sub>

peiniliu commented:
#193

terrytangyuan reviewed (commented):
Thanks for submitting the PR. Looks like you are still working on it so I changed it to WIP. Let me know when it's ready for review.

peiniliu commented:
Hi,
I think you could start reviewing.
following these examples, I find the volume is usually defined inside the workflow. 
[volume empty](https://github.com/argoproj/argo-workflows/blob/master/examples/volumes-emptydir.yaml)
[volume existing](https://github.com/argoproj/argo-workflows/blob/master/examples/volumes-existing.yaml)
[volume pvc](https://github.com/argoproj/argo-workflows/blob/master/examples/volumes-pvc.yaml)

Changes I did in couler,
(1) deleted the automatic volume definition inside container.py
(2) add get_volumeMounts_name function to get the volumeMount name of a container
(3) inside the workflow, check if the volume exists before adding the template.dic. If the volume not exists, add empty_dir

I don't know if it is somehow fits your idea, also if you have any comments on this.

terrytangyuan commented on couler/tests/volume_test.py:
```diff
@@ -0,0 +1,17 @@
+import os
```
This looks like an integration test so perhaps move it to integration tests folder and make sure to test this on CI as well similar to others.

peiniliu commented:
Hi Terry,

Sorry for the late response, Thanks for your comments! 
I have fixed all the comments you provided and also added an integration test. Let me know if more things need to be improved.

Best,
Peini

terrytangyuan commented on integration_tests/volume_example.py:
```diff
@@ -0,0 +1,59 @@
+import os
```
This is not executed on CI yet. You need to add it to https://github.com/couler-proj/couler/blob/master/scripts/integration_tests.sh.

terrytangyuan commented on couler/core/templates/workflow.py:
```diff
@@ -158,6 +166,21 @@ def to_dict(self):
+            if volume_mounts is not None:
+                for volume_mount in volume_mounts:
+                    if self.has_pvc_template(volume_mount.name) is False and self.has_volume(volume_mount.name) is False:
+                        #autogenerate emptydir volume
```
```suggestion
                        # Auto-generate emptyDir volume
```

terrytangyuan commented:
Please address unsolved comments and fix failed CI builds. Thanks.

peiniliu commented:
Hi Terry,

I have fixed the comments, the unit test from my side passed. CI failed because of the exit code == 1 in my test, which is not happening in my env, maybe I miss some CI configuration in couler?
![image](https://user-images.githubusercontent.com/24493305/118725399-314cf480-b830-11eb-80a3-8cda928e9f21.png)


Best,

Peini

terrytangyuan commented:
The sanity check failed https://github.com/couler-proj/couler/runs/2614334892

Please check out the instruction here on how to fix it: https://github.com/couler-proj/couler/blob/master/CONTRIBUTING.md#run-sanity-checks

terrytangyuan commented on couler/tests/test_data/output_golden_1.yaml:
```diff
@@ -51,3 +48,6 @@ spec:
           - cowsay
         args:
           - "{{inputs.parameters.para-consumer-0}}"
+  volumes:
```
This change doesn't look correct? It should be part of the template? (similar in other YAML files)

terrytangyuan commented on go.sum:
```diff
@@ -741,6 +748,10 @@ golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd/go.mod h1:h1NjWce9XRLGQEsW7w
 golang.org/x/sys v0.0.0-20201112073958-5cba982894dd h1:5CtCZbICpIOFdgO940moixOPjc0178IU44m4EjOO5IY=
 golang.org/x/sys v0.0.0-20201112073958-5cba982894dd/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
+golang.org/x/sys v0.0.0-20201119102817-f84b799fce68/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
+golang.org/x/sys v0.0.0-20210330210617-4fbd30eecc44/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
```
I wonder what introduced these changes.

peiniliu replied:
Hi，
I think it because I was trying to run integrated tests locally. But I think my go env(proto3) has some version problems.
I have removed these changes.

terrytangyuan commented on integration_tests/volume_example.py:
```diff
@@ -0,0 +1,72 @@
+            ],
+            volume_mounts=[VolumeMount("apppath", "/tmp")],
+        )
+        # 3) Add an exit handler that runs when the workflow succeeds.
```
Simplify this example by removing the exit handler

terrytangyuan commented on integration_tests/volume_example.py:
```diff
@@ -0,0 +1,72 @@
+                ' vol_found=`mount | grep /tmp` && \
+            if [[ -n $vol_found ]]; \
+            then echo "Volume mounted and found"; \
+            else echo "Not found"; fi '
```
How do we know if this is correct? Should we be exiting 0 and non-zero here instead of just echo-ing here?

peiniliu replied:
yes, I removed the handler, and add exit -1 instead of not found.

terrytangyuan commented:
@peiniliu Thanks for fixing the tests. I previous overlooked some of the details and just left some additional (minor) comments. Please take another look.

terrytangyuan reviewed (approved):
Thank you!
